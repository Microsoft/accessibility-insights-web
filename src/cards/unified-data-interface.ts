// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

// this is similar to `TestEngine` interface from axe-core
interface ScanEngine {
    name: string;
    version: string;
}

interface OSData {
    name: string;
    osVersion?: string;
    userAgent?: string;
}

interface ViewPortData {
    platformHeight?: number;
    platformWidth?: number;
}

interface PlatformData {
    osInfo: OSData;
    viewPortInfo: ViewPortData;
}

interface ToolData {
    scanEngine: ScanEngine;
}

interface ScanMetaData {}

interface ScanResults {
    results: UnifiedResults;
    platformInfo: PlatformData;
    toolInfo: ToolData;
    scanMetaData: ScanMetaData;
}

interface UnifiedResults {
    passes: UnifiedResultInstance[];
    violations: UnifiedResultInstance[];
    incomplete: UnifiedResultInstance[];
}

interface UnifiedRuleResult {
    id: string;
    ruleDescription: string;
}

interface InstancePropertyBag<T> {
    [property: string]: T;
}

type StoredInstancePropertyBag = InstancePropertyBag<any>;

interface UnifiedResultInstance {
    id: string;
    evaluation: StoredInstancePropertyBag;
    identifiers: StoredInstancePropertyBag;
    descriptors: StoredInstancePropertyBag;
    resolution: StoredInstancePropertyBag;
}
