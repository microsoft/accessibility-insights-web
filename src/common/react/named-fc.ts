// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';

export type ReactFCWithDisplayName<P = {}> = React.FC<React.PropsWithChildren<P>> & {
    displayName: string;
};

export function NamedFC<P = {}>(
    displayName: string,
    component: React.FC<React.PropsWithChildren<P>>,
): ReactFCWithDisplayName<P> {
    component.displayName = displayName;

    return component as ReactFCWithDisplayName<P>;
}
