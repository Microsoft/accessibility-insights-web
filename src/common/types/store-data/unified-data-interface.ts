// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { GuidanceLink } from '../../../scanner/rule-to-links-mappings';

// this is similar to `TestEngine` interface from axe-core
export interface ScanEngineProperties {
    name: string;
    version: string;
}

export interface ApplicationProperties {
    name: string;
    version: string;
}

export interface OSProperties {
    name: string;
    version: string;
    userAgent?: string;
}

export interface ViewPortProperties {
    height?: number;
    width?: number;
}

export interface PlatformData {
    osInfo: OSProperties;
    viewPortInfo: ViewPortProperties;
}

export interface TargetAppData {
    name: string;
    url?: string;
}

export interface ToolData {
    scanEngineProperties: ScanEngineProperties;
    applicationProperties: ApplicationProperties;
}

export interface UnifiedRule {
    id: string;
    description: string;
    url: string;
    guidance: GuidanceLink[];
}

export interface UnifiedScanResultStoreData {
    results: UnifiedResult[];
    rules: UnifiedRule[];
    platformInfo?: PlatformData;
    toolInfo?: ToolData;
    targetAppInfo?: TargetAppData;
}

export interface InstancePropertyBag {
    [property: string]: any;
}

export type StoredInstancePropertyBag = InstancePropertyBag;

export type UnifiedIdentifiers = {
    identifier: string;
    conciseName: string;
} & InstancePropertyBag;

export type UnifiedDescriptors = {
    snippet?: string;
} & InstancePropertyBag;

export type UnifiedResolution = {
    howToFixSummary: string;
} & InstancePropertyBag;

export interface UnifiedResult {
    uid: string;
    status: InstanceResultStatus;
    ruleId: string;
    identifiers: UnifiedIdentifiers;
    descriptors: UnifiedDescriptors;
    resolution: UnifiedResolution;
}

export type InstanceResultStatus = 'pass' | 'fail' | 'unknown';

export type UnifiedRuleResultStatus = InstanceResultStatus | 'inapplicable';

export interface UnifiedRuleResult {
    id: string;
    nodes: UnifiedResult[];
    description: string;
    url: string;
    guidance: GuidanceLink[];
}

export type UnifiedStatusResults = {
    [key in UnifiedRuleResultStatus]: UnifiedRuleResult[];
};

export const AllRuleResultStatuses: UnifiedRuleResultStatus[] = ['pass', 'fail', 'unknown', 'inapplicable'];
