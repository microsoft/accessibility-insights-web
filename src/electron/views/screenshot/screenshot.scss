// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
@import '../../../common/styles/colors.scss';

.screenshot-image {
    width: 100%;
    object-fit: contain;
    outline: 1px solid $screenshot-image-outline;
    box-shadow: 0 6px 12px $neutral-alpha-20;

    &:focus {
        outline-offset: 4px;
        outline: 2px solid $primary-text;
    }
}
