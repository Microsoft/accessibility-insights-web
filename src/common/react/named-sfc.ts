// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';

export type ReactSFCWithDisplayName<P = {}> = React.SFC<P> & { displayName: string };

export function NamedSFC<P = {}>(displayName: string, component: React.SFC<P>): ReactSFCWithDisplayName<P> {
    component.displayName = displayName;

    return component as ReactSFCWithDisplayName<P>;
}
