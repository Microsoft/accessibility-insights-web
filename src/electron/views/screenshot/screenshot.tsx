// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { NamedFC } from 'common/react/named-fc';
import * as React from 'react';

export type ScreenshotProps = {
    encodedImage: string;
    altText: string;
};

export const Screenshot = NamedFC<ScreenshotProps>('Screenshot', props => {
    return <img src={'data:image/png;base64,' + props.encodedImage} alt={props.altText} />;
});
