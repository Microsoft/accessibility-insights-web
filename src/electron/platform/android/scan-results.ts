// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { ScreenshotData } from 'common/types/store-data/unified-data-interface';

export interface RuleResultsData {
    axeViewId: string;
    ruleId: string;
    status: string;
    props: any;
}

export interface ViewElementData {
    axeViewId: string;
    boundsInScreen: BoundingRectangle;
    className: string;
    contentDescription: string;
    text: string;
    children: ViewElementData[];
}

export interface BoundingRectangle {
    bottom: number;
    left: number;
    right: number;
    top: number;
}

export class ScanResults {
    constructor(readonly rawData: any) {}

    public get deviceName(): string {
        try {
            return this.rawData.axeContext.axeDevice.name;
        } catch {
            return null;
        }
    }

    public get appIdentifier(): string {
        try {
            return this.rawData.axeContext.axeMetaData.appIdentifier;
        } catch {
            return null;
        }
    }

    public get viewElementTree(): ViewElementData {
        try {
            return this.rawData.axeContext.axeView;
        } catch {
            return null;
        }
    }

    public get ruleResults(): RuleResultsData[] {
        try {
            const results = this.rawData.axeRuleResults;
            return results || [];
        } catch {
            return [];
        }
    }

    public get axeVersion(): string {
        try {
            return this.rawData.axeContext.axeMetaData.axeVersion;
        } catch {
            return 'no-version';
        }
    }

    public get screenshot(): ScreenshotData {
        try {
            return {
                base64PngData: this.rawData.axeContext.screenshot,
            };
        } catch {
            return null;
        }
    }
}
