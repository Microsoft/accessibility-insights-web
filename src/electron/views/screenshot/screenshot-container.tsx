// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { NamedFC } from 'common/react/named-fc';
import { ScreenshotData } from 'common/types/store-data/unified-data-interface';
import { BoundingRectangle } from 'electron/platform/android/scan-results';
import { HighlightBox } from 'electron/views/screenshot/highlight-box';
import { Screenshot } from 'electron/views/screenshot/screenshot';
import { screenshotContainer } from 'electron/views/screenshot/screenshot-container.scss';
import { isEmpty } from 'lodash';
import * as React from 'react';

export interface ScreenshotContainerProps {
    screenshotData?: ScreenshotData;
    highlightBoxRectangles: BoundingRectangle[];
}

export const ScreenshotContainer = NamedFC<ScreenshotContainerProps>('ScreenshotContainer', props => {
    if (isEmpty(props.screenshotData)) {
        return <div>Screenshot for scan is unavailable</div>;
    }

    return (
        <div className={screenshotContainer}>
            <Screenshot encodedImage={props.screenshotData.base64PngData} />
            {renderHighlightBoxes(props.highlightBoxRectangles)}
        </div>
    );
});

function renderHighlightBoxes(highlightBoxRectangles: BoundingRectangle[]): JSX.Element[] {
    if (isEmpty(highlightBoxRectangles)) {
        return;
    }

    const renderedBoxes: JSX.Element[] = [];

    // When making the highlight boxes interactive, it would be better to use the result uid as the key, so
    // the mapping between results and highlight boxes is maintained if boxes are added/removed from the list.
    highlightBoxRectangles.forEach((rectangle, index) => {
        renderedBoxes.push(<HighlightBox key={index} boundingRectangle={rectangle} />);
    });

    return renderedBoxes;
}
