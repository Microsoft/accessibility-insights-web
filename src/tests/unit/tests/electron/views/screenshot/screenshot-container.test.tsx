// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { ScreenshotData } from 'common/types/store-data/unified-data-interface';
import { BoundingRectangle } from 'electron/platform/android/scan-results';
import { ScreenshotContainer, ScreenshotContainerProps } from 'electron/views/screenshot/screenshot-container';
import { shallow } from 'enzyme';
import * as React from 'react';

describe('ScreenshotContainer', () => {
    const basicScreenshotData: ScreenshotData = { base64PngData: 'test-base-64-png-data' };

    describe('screenshot', () => {
        it('renders when passed a value for screenshotData', () => {
            const wrapper = shallow(<ScreenshotContainer screenshotData={basicScreenshotData} highlightBoxRectangles={[]} />);

            expect(wrapper.getElement()).toMatchSnapshot();
        });

        const emptyScreenshotDataCases: ScreenshotData[] = [null, undefined, {} as ScreenshotData];

        it.each(emptyScreenshotDataCases)(
            'renders screenshot unavailable string when passed empty screenshotData %p',
            (screenshotDataCase: ScreenshotData) => {
                const wrapper = shallow(<ScreenshotContainer screenshotData={screenshotDataCase} highlightBoxRectangles={[]} />);

                expect(wrapper.getElement()).toMatchSnapshot();
            },
        );
    });

    describe('highlight boxes', () => {
        it('do not render when highlightBoxRectangles array is empty', () => {
            const props: ScreenshotContainerProps = {
                screenshotData: basicScreenshotData,
                highlightBoxRectangles: [],
            };

            const wrapper = shallow(<ScreenshotContainer {...props} />);

            expect(wrapper.getElement()).toMatchSnapshot();
        });

        it('render when passed values for highlightBoxRectangles array', () => {
            const highlightBoxRectangles: BoundingRectangle[] = [
                { top: 0, bottom: 100, left: 0, right: 100 },
                { top: 150, bottom: 200, left: 150, right: 300 },
            ];

            const props: ScreenshotContainerProps = {
                screenshotData: basicScreenshotData,
                highlightBoxRectangles: highlightBoxRectangles,
            };

            const wrapper = shallow(<ScreenshotContainer {...props} />);

            expect(wrapper.getElement()).toMatchSnapshot();
        });
    });
});
