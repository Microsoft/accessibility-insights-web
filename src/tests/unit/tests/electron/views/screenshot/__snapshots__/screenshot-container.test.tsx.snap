// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScreenshotContainer highlight boxes do not render when highlightBoxRectangles array is empty 1`] = `
<div
  className="screenshotContainer"
>
  <Screenshot
    encodedImage="test-base-64-png-data"
  />
</div>
`;

exports[`ScreenshotContainer highlight boxes render when passed values for highlightBoxRectangles array 1`] = `
<div
  className="screenshotContainer"
>
  <Screenshot
    encodedImage="test-base-64-png-data"
  />
  <HighlightBox
    boundingRectangle={
      Object {
        "bottom": 100,
        "left": 0,
        "right": 100,
        "top": 0,
      }
    }
  />
  <HighlightBox
    boundingRectangle={
      Object {
        "bottom": 200,
        "left": 150,
        "right": 300,
        "top": 150,
      }
    }
  />
</div>
`;

exports[`ScreenshotContainer screenshot renders screenshot unavailable string when passed empty screenshotData {} 1`] = `
<div>
  Screenshot for scan is unavailable
</div>
`;

exports[`ScreenshotContainer screenshot renders screenshot unavailable string when passed empty screenshotData null 1`] = `
<div>
  Screenshot for scan is unavailable
</div>
`;

exports[`ScreenshotContainer screenshot renders screenshot unavailable string when passed empty screenshotData undefined 1`] = `
<div>
  Screenshot for scan is unavailable
</div>
`;

exports[`ScreenshotContainer screenshot renders when passed a value for screenshotData 1`] = `
<div
  className="screenshotContainer"
>
  <Screenshot
    encodedImage="test-base-64-png-data"
  />
</div>
`;
